dlt/common/libs/pyarrow.py

```diff
@@ -634,124 +649,151 @@ def concat_batches_and_tables_in_order(
             )
-            columnar_known_types[field.name] = json_str_array
-
-    # If there are unknown type columns, first create a table to infer their types
-    if columnar_unknown_types:
-        new_schema_fields = []
-        for key in list(columnar_unknown_types):
-            arrow_col: Optional[pa.Array] = None
-            try:
-                arrow_col = pa.array(columnar_unknown_types[key])
-                if pa.types.is_null(arrow_col.type):
-                    logger.warning(
-                        f"Column {key} contains only NULL values and data type could not be"
-                        " inferred. This column is removed from a arrow table"
-                    )
-                    continue
-
-            except pa.ArrowInvalid as e:
-                # Try coercing types not supported by arrow to a json friendly format
-                # E.g. dataclasses -> dict, UUID -> str
-                try:
-                    arrow_col = pa.array(
-                        map_nested_in_place(custom_encode, list(columnar_unknown_types[key]))
-                    )
-                    logger.warning(
-                        f"Column {key} contains a data type which is not supported by pyarrow and"
-                        f" got converted into {arrow_col.type}. This slows down arrow table"
-                        " generation."
-                    )
-                except (pa.ArrowInvalid, TypeError):
-                    logger.warning(
-                        f"Column {key} contains a data type which is not supported by pyarrow. This"
-                        f" column will be ignored. Error: {e}"
-                    )
-            if arrow_col is not None:
-                columnar_known_types[key] = arrow_col
-                new_schema_fields.append(
-                    pa.field(
-                        key,
-                        arrow_col.type,
-                        nullable=columns[key].get("nullable", True),
-                    )
+        except (pa.ArrowInvalid, TypeError):
+            logger.warning("Couldn't successfully encode Python objects to string and create pyarrow array.")
+
+    # case 2: if `base case` failed, the array contains non-string objects (likely dict, tuple, list, set, etc.)
+    # encode these objects as JSON strings; 
+    if arrow_array is None and dlt_data_type == "json":
+        logger.warning(
+            "Received `data_type='json'`, data requries serialization to string, slowing extraction. "
+            "Cast the JSON field to STRING in your database system to improve performance. "
+            "For example, create and extract data from an SQL VIEW that SELECT with CAST."
+        )
+        arrow_array = pa.array(
+            [
+                (
+                    None if item is None
+                    else json_to_str(list(item)) if issubclass(type(item), set)  # sets are not supported by json_to_str
```

Comment:
this function should be called `typed_json_to_str` because before dumping it removes type encoding from str values. due to that it will be pretty slow. my take would be to use `json` directly, you can also experiment with `dumpb` and casting to str in arrow. this will be the fastest, we use `orjson` by default and it dumps to binary natively